File: infra/web-adapter/src/main/kotlin/com/wallet/clover/adapter/JsoupTicketParser.kt
package com.wallet.clover.adapter

import com.wallet.clover.domain.game.LottoGameStatus
import com.wallet.clover.domain.ticket.LottoTicketStatus
import com.wallet.clover.domain.ticket.parser.ParsedGame
import com.wallet.clover.domain.ticket.parser.ParsedTicket
import com.wallet.clover.domain.ticket.parser.TicketParser
import com.wallet.clover.domain.ticket.parser.TicketParsingException
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class JsoupTicketParser : TicketParser {
    private val logger = LoggerFactory.getLogger(javaClass)

    override fun parse(html: String): ParsedTicket {
        val document = Jsoup.parse(html)
        val ordinal = getOrdinal(document)
        val status = getTicketStatus(document)
        val games = getGames(document)
        return ParsedTicket(ordinal, status, games)
    }

    private fun getOrdinal(document: Document): Int {
        val content = document.select("h3 > span.key_clr1").firstOrNull()?.text() ?: "0"
        logger.debug("Parsed ordinal content: {}", content)
        return content.filter { it.isDigit() }.toIntOrNull() ?: 0
    }

    private fun getTicketStatus(document: Document): LottoTicketStatus {
        val content = document.select("div.bx_notice.winner strong").firstOrNull()?.text() ?: ""
        logger.debug("Parsed ticket status content: {}", content)
        return when {
            content.contains("당첨") -> LottoTicketStatus.WINNING
            content.contains("낙첨") -> LottoTicketStatus.LOSING
            content.contains("추첨") -> LottoTicketStatus.STASHED // "미추첨" 대신 "추첨" 키워드 사용
            else -> {
                logger.error("Unidentifiable ticket status content: {}", content)
                throw TicketParsingException("'$content' 는 식별할 수 없는 문구 입니다.")
            }
        }
    }

    private fun getGames(document: Document): List<ParsedGame> {
        logger.info("Parsing games from document")
        val gameRows = document.select("div.list_my_number table tbody tr")
        return gameRows.map { row ->
            val resultText = row.select("td.result").text().trim()
            val numbers = row.select("td span.clr").mapNotNull { it.text().toIntOrNull() }

            if (numbers.size != 6) {
                logger.error("Parsed game has incorrect number count: {}", numbers)
                throw TicketParsingException("게임의 번호가 6개가 아닙니다.")
            }

            ParsedGame(
                status = LottoGameStatus.valueOfHtmlValue(resultText),
                number1 = numbers[0],
                number2 = numbers[1],
                number3 = numbers[2],
                number4 = numbers[3],
                number5 = numbers[4],
                number6 = numbers[5],
            )
        }.also {
            logger.info("Successfully parsed {} games.", it.size)
        }
    }
}

File: apps/backend/src/ticket/client/jsoup-ticket.parser.ts
import { Injectable, Logger, BadRequestException } from '@nestjs/common';
import * as cheerio from 'cheerio';

/**
 * 파싱된 티켓 정보 인터페이스
 */
export interface ParsedTicket {
  ordinal: number;
  status: string;
  games: ParsedGame[];
}

/**
 * 파싱된 개별 게임 정보 인터페이스
 */
export interface ParsedGame {
  status: string;
  number1: number;
  number2: number;
  number3: number;
  number4: number;
  number5: number;
  number6: number;
}

/**
 * 동행복권 HTML 문서를 파싱하여 로또 데이터를 추출하는 클래스입니다.
 */
@Injectable()
export class JsoupTicketParser {
  private readonly logger = new Logger(JsoupTicketParser.name);

  // 셀렉터 정의
  private readonly selectors = {
    ordinalSelector: 'h3 > span.key_clr1',
    ticketStatusSelector: 'div.bx_notice.winner strong',
    gameRowsSelector: 'div.list_my_number table tbody tr',
    gameResultSelector: 'td.result',
    gameNumbersSelector: 'td span.clr',
  };

  parse(html: string): ParsedTicket {
    const $ = cheerio.load(html);

    const ordinal = this.getOrdinal($);
    const status = this.getTicketStatus($);
    const games = this.getGames($);

    return { ordinal, status, games };
  }

  private getOrdinal($: cheerio.CheerioAPI): number {
    const text = $(this.selectors.ordinalSelector).first().text() || '0';
    return parseInt(text.replace(/[^0-9]/g, ''), 10) || 0;
  }

  private getTicketStatus($: cheerio.CheerioAPI): string {
    const text = $(this.selectors.ticketStatusSelector).first().text() || '';

    if (text.includes('당첨')) return 'WINNING';
    if (text.includes('낙첨')) return 'LOSING';
    if (text.includes('추첨')) return 'STASHED';

    this.logger.error(`식별할 수 없는 티켓 상태: ${text}`);
    throw new BadRequestException(`식별할 수 없는 티켓 상태입니다: ${text}`);
  }

  private getGames($: cheerio.CheerioAPI): ParsedGame[] {
    const games: ParsedGame[] = [];
    const rows = $(this.selectors.gameRowsSelector);

    rows.each((_, element) => {
      const game = this.parseGameRow($, element);
      if (game) games.push(game);
    });

    return games;
  }

  private parseGameRow($: cheerio.CheerioAPI, element: any): ParsedGame | null {
    const row = $(element);
    const resultText = row
      .find(this.selectors.gameResultSelector)
      .text()
      .trim();
    const numbers = row
      .find(this.selectors.gameNumbersSelector)
      .map((_, el) => parseInt($(el).text(), 10))
      .get();

    if (numbers.length !== 6) return null;

    return {
      status: this.parseGameStatus(resultText),
      number1: numbers[0],
      number2: numbers[1],
      number3: numbers[2],
      number4: numbers[3],
      number5: numbers[4],
      number6: numbers[5],
    };
  }

  private parseGameStatus(text: string): string {
    const statusMap: { [key: string]: string } = {
      '1등당첨': 'WINNING_1',
      '2등당첨': 'WINNING_2',
      '3등당첨': 'WINNING_3',
      '4등당첨': 'WINNING_4',
      '5등당첨': 'WINNING_5',
    };

    for (const [key, value] of Object.entries(statusMap)) {
      if (text.includes(key)) return value;
    }

    return 'LOSING';
  }
}
